Launch Sites Locations Analysis with Folium
Estimated time needed: 40 minutes

The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.

In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using matplotlib and seaborn and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using Folium.

Objectives
This lab contains the following tasks:

TASK 1: Mark all launch sites on a map
TASK 2: Mark the success/failed launches for each site on the map
TASK 3: Calculate the distances between a launch site to its proximities
After completed the above tasks, you should be able to find some geographical patterns about launch sites.

Let's first import required Python packages for this lab:

In [1]:
!pip3 install folium
!pip3 install wget

  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9672 sha256=d76ea54bd7d3dc76c4efc84cf73d206a5713f1cb39164b3ee07a87952064350c
  Stored in directory: c:\users\linyi\appdata\local\pip\cache\wheels\04\5f\3e\46cc37c5d698415694d83f607f833f83f0149e49b3af9d0f38
Successfully built wget


In [2]:
import folium
import wget
import pandas as pd

In [ ]:
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

Task 1: Mark all launch sites on a map
First, let's try to add each site's location on a map using site's latitude and longitude coordinates

The following dataset with the name spacex_launch_geo.csv is 

In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

100% [................................................................................] 7710 / 7710

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.

We first need to create a folium Map object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.

In [5]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

In [6]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

Now, you can explore the map by zoom-in/out the marked areas , and try to answer the following questions:

Are all launch sites in proximity to the Equator line?
Are all launch sites in very close proximity to the coast?
Also please try to explain your findings.

Explanation:
Most of Launch sites considered in this project are in proximity to the Equator line. Launch sites are made at the closest point possible to Equator line, because anything on the surface of the Earth at the equator is already moving at the maximum speed (1670 kilometers per hour). For example launching from the equator makes the spacecraft move almost 500 km/hour faster once it is launched compared half way to north pole.
All launch sites considered in this project are in very close proximity to the coast While starting rockets towards the ocean we minimise the risk of having any debris dropping or exploding near people.

Task 2: Mark the success/failed launches for each site on the map
Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates. Recall that data frame spacex_df has detailed launch records, and the class column indicates if this launch was successful or not

In [7]:
spacex_df.tail(10)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
46                         SES-11 / EchoStar 105            5200.00   
47                                   Koreasat 5A            3500.00   
48                                 Bangabandhu-1            3600.00   
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
46        GTO         SES EchoStar  Success (drone ship)      1  28.573255   
47        GTO       KT Corporation  Success (drone ship)      1  28.573255   
48        GTO   Thales-Alenia/BTRC  Success (drone ship)      1  28.573255   
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

         Long  
46 -80.646895  
47 -80.646895  
48 -80.646895  
49 -80.576820  
50 -80.576820  
51 -80.576820  
52 -80.576820  
53 -80.576820  
54 -80.576820  
55 -80.576820

Next, let's create markers for all launch records. If a launch was successful (class=1), then we use a green marker and if a launch was failed, we use a red marker (class=0)

Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.

Let's first create a MarkerCluster object

In [8]:
marker_cluster = MarkerCluster()

In [9]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
spacex_df['marker_color'] = list(map(lambda x: 'green' if x==1 else 'red', spacex_df['class']))
spacex_df.tail(10)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
46                         SES-11 / EchoStar 105            5200.00   
47                                   Koreasat 5A            3500.00   
48                                 Bangabandhu-1            3600.00   
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
46        GTO         SES EchoStar  Success (drone ship)      1  28.573255   
47        GTO       KT Corporation  Success (drone ship)      1  28.573255   
48        GTO   Thales-Alenia/BTRC  Success (drone ship)      1  28.573255   
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

         Long marker_color  
46 -80.646895        green  
47 -80.646895        green  
48 -80.646895        green  
49 -80.576820        green  
50 -80.576820        green  
51 -80.576820          red  
52 -80.576820          red  
53 -80.576820          red  
54 -80.576820        green  
55 -80.576820          red

In [10]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
46                         SES-11 / EchoStar 105            5200.00   
47                                   Koreasat 5A            3500.00   
48                                 Bangabandhu-1            3600.00   
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
46        GTO         SES EchoStar  Success (drone ship)      1  28.573255   
47        GTO       KT Corporation  Success (drone ship)      1  28.573255   
48        GTO   Thales-Alenia/BTRC  Success (drone ship)      1  28.573255   
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

         Long marker_color  
46 -80.646895        green  
47 -80.646895        green  
48 -80.646895        green  
49 -80.576820        green  
50 -80.576820        green  
51 -80.576820          red  
52 -80.576820          red  
53 -80.576820          red  
54 -80.576820        green  
55 -80.576820          red

In [11]:
# Add the Marker cluster to the site map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for site_lat, site_long, marker_color in zip(spacex_df['Lat'], spacex_df['Long'], spacex_df['marker_color']):
    site_coordinate = [site_lat, site_long]
    marker = folium.map.Marker(
        site_coordinate,
        # Create an icon as a text label
        icon=folium.Icon(color='white', 
                         icon_color=marker_color)
    )
    marker.add_to(marker_cluster)

site_map

TASK 3: Calculate the distances between a launch site to its proximities
Next, we need to explore and analyze the proximities of launch sites.

Let's first add a MousePosition on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)



In [13]:
 #Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.

You can calculate the distance between two points on the map based on their Lat and Long values using the following method:

In [18]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

In [20]:
# distance_railway = calculate_distance(lat1, lon1, lat2, lon2)
railway_marker = [28.55752, -80.80155]
launch_coordinate = [28.57337, -80.64669]
distance_railway = calculate_distance(railway_marker[0], railway_marker[1], launch_coordinate[0], launch_coordinate[1])
distance_railway  # distance in km

15.230651245141603

In [21]:
# create and add a folium.Marker on your selected closest raiwaly point on the map
# show the distance to the launch site using the icon property 
marker = folium.map.Marker(
        railway_marker,
        # Create an icon as a text label
        icon=DivIcon(
            icon_size=(400, 400),
            icon_anchor=(0, 0),
            html='<div style="font-size:400; color:#0c10f2;"><b>%s</b></div>' % str(round(distance_railway, 2))+' km',
            )
    )
marker.add_to(site_map)

site_map

In [23]:
 #Create a marker with distance to a closest city, coastline, highway, etc.
# Draw a line between the marker to the launch site
city   = [28.61200, -80.80788]
coastline = [28.5858, -80.79952]
highway   = [28.5402, -80.85079]

city_distance = calculate_distance(city[0], city[1], launch_coordinate[0], launch_coordinate[1])
coastline_distance = calculate_distance(coastline[0], coastline[1], launch_coordinate[0], launch_coordinate[1])
highway_distance = calculate_distance(highway[0], highway[1], launch_coordinate[0], launch_coordinate[1])

colors = ['red','orange','green']
html_colors = ['#dc3545','#fd7e14','#198754']

for coordinate ,distance, color, html_color in zip([city, coastline, highway], [city_distance, coastline_distance, highway_distance], colors, html_colors):
    marker = folium.map.Marker(
            coordinate,
            # Create an icon as a text label
            icon=DivIcon(
                icon_size=(20,20),
                icon_anchor=(0,0),
                html='<div style="font-size: 12; color:'+html_color+';"><b>%s</b></div>' % str(round(distance, 2)) + 'km',
                )
            )
    marker.add_to(site_map)
    folium.PolyLine([coordinate, launch_coordinate], color=color).add_to(site_map)
site_map

After you plot distance lines to the proximities, you can answer the following questions easily:

Are launch sites in close proximity to railways?
Are launch sites in close proximity to highways?
Are launch sites in close proximity to coastline?
Do launch sites keep certain distance away from cities?
Also please try to explain your findings.

Explanation:
From the visual analysis of the launch site KSC LC-39A we can clearly see that it is:
relative close to railway (15.23 km)
relative close to highway (20.28 km)
relative close to coastline (14.99 km)
Also the launch site KSC LC-39A is relative close to its closest city Titusville (16.32 km).
Failed rocket with its high speed can cover distances like 15-20 km in few seconds. It could be potentially dangerous to populated areas.

Authors
Yan Luo